File: 19-remove-nth-node-from-end-of-list/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# class Solution(object):
#     def removeNthFromEnd(self, head, n):
#         """
#         :type head: ListNode
#         :type n: int
#         :rtype: ListNode
#         """
#         curr, p = head, head
        
#         count = 0
#         while curr != None:
#             count += 1
#             curr = curr.next
        
#         i = 1
#         pre = None
#         while p != None:
#             if i == count - n + 1:
#                 if pre != None:
#                     pre.next = p.next
#                 else:
#                     head = head.next
#                 return head
#             i += 1
#             pre = p
#             p = p.next
    
#         return head

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        L = 0
        curr = head
        while curr != None:
            L += 1
            curr = curr.next
        
        L = L-n
        dummy = ListNode(sys.maxint)
        dummy.next = head
        curr = dummy
        
        while L > 0:
            L -= 1
            curr = curr.next
        curr.next = curr.next.next
        return dummy.next
        

File: 19-remove-nth-node-from-end-of-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* node = findNthNodeFromEnd(dummy, n + 1);
        ListNode* p = node->next;
        node->next = node->next->next;
        delete p;
        return dummy->next;
    }
private:
    ListNode* findNthNodeFromEnd(ListNode* head, int n) {
        ListNode* p1 = head;
        for (int i = 0; i < n; i ++) {
            p1 = p1->next;
            if (!p1) break;
        }
        ListNode * p2 = head;
        while (p1) {
            p1 = p1->next;
            p2 = p2->next;
        }
        return p2;
    }
};
